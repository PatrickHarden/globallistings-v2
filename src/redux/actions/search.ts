import { createStandardAction } from 'typesafe-actions';
import { Dictionary } from '../../types/common/dictionary';

export const setSearchLocation = createStandardAction('SET_SEARCH_LOCATION')<string>();
export const changeSearchFilter = createStandardAction('CHANGE_SEARCH_FILTER')<
  Dictionary<string>
>();
