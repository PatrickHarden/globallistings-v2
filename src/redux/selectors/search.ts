// @ts-nocheck
import { createSelector } from 'reselect';
import { stringify } from 'querystring';

import { configSelector, filtersSelector } from './state';
import { intlSelector } from './intl';
import {
  searchPathSelector as uiSearchPathSelector,
  suggestionsSelector,
  selectedSuggestionSelector,
  selectedFiltersSelector,
} from './ui';
import {
  hideBuyButtonSelector,
  hideLetButtonSelector,
  searchResultsPageSelector,
  searchPathSelectorSelector as configSearchPathSelectorSelector,
} from './config/config';
import { defaultFiltersParamsSelectors } from './config/filters';
import { TransactionTypeValue } from '../../types/config';
import { plpUrlPathParamsSelector } from './match';
import { filterParamsCombiner } from './config/range-filters';

export interface ButtonInfo {
  readonly text: string;
  readonly url: string;
  readonly visible: boolean;
}

export const buyButtonTextSelector = createSelector(
  hideBuyButtonSelector,
  intlSelector,
  (hideBuyButton, intl) =>
    hideBuyButton
      ? intl.formatMessage({ id: 'GenericSearchButton' })
      : intl.formatMessage({ id: 'SearchToBuyButton' })
);

export const letButtonTextSelector = createSelector(
  hideBuyButtonSelector,
  intlSelector,
  (hideBuyButton, intl) =>
    hideBuyButton
      ? intl.formatMessage({ id: 'GenericSearchButton' })
      : intl.formatMessage({ id: 'SearchToLetButton' })
);

export const displaySearchPathSelectorSelector = createSelector(
  configSelector,
  config =>
    config.searchConfig.searchPathSelector && config.searchConfig.searchPathSelector.length > 0
);

const defaultSearchPathValueSelector = createSelector(
  configSearchPathSelectorSelector,
  searchPaths => {
    if (!searchPaths || !searchPaths.length) {
      return undefined;
    }

    const defaultSearchPath = searchPaths.find(sp => sp.default);

    if (!defaultSearchPath) {
      throw new Error('If searchPaths are present there must always be a default one');
    }

    return defaultSearchPath.value;
  }
);

export const searchPathSelector = createSelector(
  uiSearchPathSelector,
  defaultSearchPathValueSelector,
  (uiSearchPath, defaultSearchPath) => uiSearchPath || defaultSearchPath
);

export const plpPathSelector = createSelector(
  searchPathSelector,
  searchResultsPageSelector,
  (searchPath, searchResultsPage) => {
    if (!searchPath && !searchResultsPage) {
      throw new Error('SearchPath of search');
    }

    return searchPath || searchResultsPage;
  }
);

const filtersParamsSelector = createSelector(
  selectedFiltersSelector,
  defaultFiltersParamsSelectors,
  (selectedFilters, defaultFilters) => ({ ...defaultFilters, ...selectedFilters })
);

export const v1FilterParamsSelector = createSelector(
  filtersParamsSelector,
  filtersSelector,
  () => true,
  () => false,
  filterParamsCombiner
);

const uiSearchParamsSelector = createSelector(selectedSuggestionSelector, suggestion =>
  !!suggestion ? { placeId: suggestion.placeId, location: suggestion.description } : {}
);

const searchParamsSelector = createSelector(
  v1FilterParamsSelector,
  uiSearchParamsSelector,
  (filterParams, uiParams) => ({ ...filterParams, ...uiParams })
);

export const searchLinkSelector = (linkType: TransactionTypeValue) =>
  createSelector(
    plpPathSelector,
    searchParamsSelector,
    selectedSuggestionSelector,
    (plpPath, params, suggestion) => {
      const newParams = {
        ...params,
        aspects: linkType,
      };

      // TODO: this needs to be moved into state and a selector created to retrieve it
      // also a non null assertion operator added and tested
      // TODO: do we really want to add /details here?
      if (suggestion !== undefined && suggestion!.propertyResult !== undefined) {
        if (suggestion.propertyResult!.propertyPartialPath !== undefined) {
          return `${suggestion.propertyResult!.searchPath}/details/${
            suggestion.propertyResult!.propertyPartialPath
          }?view=${linkType}`;
        }
      }

      return `${plpPath}?${stringify(newParams)}`;
    }
  );

export const buyButtonInfoSelector = createSelector(
  buyButtonTextSelector,
  searchLinkSelector(TransactionTypeValue.isSale),
  hideBuyButtonSelector,
  selectedSuggestionSelector,
  (text, url, hide, suggestion): ButtonInfo => {
    const toShow =
      suggestion !== undefined &&
      suggestion!.propertyResult !== undefined &&
      suggestion!.propertyResult!.propertyAspects !== undefined
        ? suggestion!.propertyResult!.propertyAspects!.includes('isSale')
        : !hide;
    return {
      text,
      url,
      visible: toShow,
    };
  }
);

export const letButtonInfoSelector = createSelector(
  letButtonTextSelector,
  searchLinkSelector(TransactionTypeValue.isLetting),
  hideLetButtonSelector,
  selectedSuggestionSelector,
  (text, url, hide, suggestion): ButtonInfo => {
    const toShow =
      suggestion !== undefined &&
      suggestion!.propertyResult !== undefined &&
      suggestion!.propertyResult!.propertyAspects !== undefined
        ? suggestion!.propertyResult!.propertyAspects!.includes('isLetting')
        : !hide;
    return {
      text,
      url,
      visible: toShow,
    };
  }
);

export interface AutocompleteSuggestion {
  readonly label: string;
  readonly value: string;
  readonly default: boolean;
}

export const locationLookupSuggestionsSelector = createSelector(
  suggestionsSelector,
  (suggestions): AutocompleteSuggestion[] =>
    suggestions.map(suggestion => ({
      label: suggestion.description,
      value: suggestion.placeId,
      default: false,
    }))
);

export const initialSearchLocationSelector = createSelector(
  plpUrlPathParamsSelector,
  ({ location }) => location
);
