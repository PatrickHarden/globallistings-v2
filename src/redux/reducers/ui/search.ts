import { DeepPartial } from '../../../types/common/deep-partial';
import { Search } from '../../../types/ui';
import * as actions from '../../actions/search';
import { ActionType } from 'typesafe-actions';

export type Actions = ActionType<typeof actions>;

export default (state: DeepPartial<Search> = {}, action: Actions) => {
  switch (action.type) {
    case 'SET_SEARCH_LOCATION':
      return {
        ...state,
        path: action.payload,
      };
    case 'CHANGE_SEARCH_FILTER':
      return {
        ...state,
        filters: { ...state.filters, ...action.payload },
      };
    default:
      return state;
  }
};
