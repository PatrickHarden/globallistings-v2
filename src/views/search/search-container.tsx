import React from 'react';
import { connect } from 'react-redux';
import {
  buyButtonInfoSelector,
  ButtonInfo,
  locationLookupSuggestionsSelector,
  AutocompleteSuggestion,
} from '../../redux/selectors/search';
import { letButtonInfoSelector } from '../../redux/selectors/search';
import { messagesSelector } from '../../redux/selectors/state';
import {
  searchModeSelector,
  searchBiasSelector,
  searchRadiusRestrictionSelector,
  searchCountryRestrictionSelector,
  getSearchHeader,
  searchResultsPageSelector,
  searchPathSelectorSelector,
  searchSuggestDebounceSelector,
  paramsSelector,
  searchPathSortAlphabeticalSelector,
  polygonSelector,
} from '../../redux/selectors/config/config';
import { FilterInfo } from '../../redux/selectors/config/filters';
import {
  primaryFiltersSelector,
  secondaryFiltersSelector,
} from '../../redux/selectors/view-filters';

import {
  setSearchLocation,
  changeSearchFilter
} from '../../redux/actions/search';
import { lookupPlaceByName } from '../../redux/actions/autocomplete-request';
import { lookupByPropertyId } from '../../redux/actions/property-request';

import { State } from '../../types/state';
import { SearchPath, SearchBias, SearchHeader, Params, Polygon } from '../../types/config';
import { Dictionary } from '../../types/common/dictionary';

import { Search } from './search';
import { selectedAutocompletionItem } from '../../redux/actions/autocomplete';
import { renderComplete } from '../../redux/actions/render';
import { siteTypeSelector } from '../../redux/selectors/config/config';

export interface StateProps {
  readonly params: Params;
  readonly biasRadius?: number;
  readonly buyButtonInfo: ButtonInfo;
  readonly language: Dictionary<string>;
  readonly letButtonInfo: ButtonInfo;
  readonly restrictToCountry?: string;
  readonly searchBias?: SearchBias;
  readonly searchHeader: SearchHeader;
  readonly searchMode: string;
  readonly searchResultsPage: string;
  readonly suggestions: AutocompleteSuggestion[];
  readonly searchPathSelectors?: SearchPath[];
  readonly primaryFilters: FilterInfo[];
  readonly secondaryFilters: FilterInfo[];
  readonly searchSuggestDebounce?: number;
  readonly searchPathSortAlphabeticalSelector?: boolean;
  readonly polygons?: Polygon[];
  readonly siteType: string
}

export interface DispatchProps {
  readonly changePlaceSearch: (searchTerm: string) => void;
  readonly changePropertySearch: (searchTerm: string) => void;
  readonly renderComplete: () => void;
  readonly changeSearchFilter: (key: string) => (value: string) => void;
  readonly setSearchLocation: (searchLocation: string) => void;
  readonly updateSearchPath: (searchPath: string) => void;
}

type Props = StateProps & DispatchProps;

export class SearchContainer extends React.Component<Props> {
  public componentDidMount() {
    this.props.renderComplete();
  }

  public render() {
    return <Search {...this.props} />;
  }
}

export const mapStateToProps = (state: State): StateProps => ({
  params: paramsSelector(state),
  biasRadius: searchRadiusRestrictionSelector(state),
  buyButtonInfo: buyButtonInfoSelector(state),
  language: messagesSelector(state),
  letButtonInfo: letButtonInfoSelector(state),
  restrictToCountry: searchCountryRestrictionSelector(state),
  searchBias: searchBiasSelector(state),
  searchHeader: getSearchHeader(state),
  searchMode: searchModeSelector(state),
  searchResultsPage: searchResultsPageSelector(state),
  suggestions: locationLookupSuggestionsSelector(state),
  searchPathSelectors: searchPathSelectorSelector(state),
  primaryFilters: primaryFiltersSelector(state),
  secondaryFilters: secondaryFiltersSelector(state),
  searchSuggestDebounce: searchSuggestDebounceSelector(state),
  searchPathSortAlphabeticalSelector: searchPathSortAlphabeticalSelector(state),
  polygons: polygonSelector(state),
  siteType: siteTypeSelector(state)
});

export const mapDispatchToProps = (dispatch: Function): DispatchProps => ({
  changePlaceSearch: searchTerm => dispatch(lookupPlaceByName(searchTerm)),
  changePropertySearch: searchTerm => dispatch(lookupByPropertyId(searchTerm)),
  changeSearchFilter: key => value => dispatch(changeSearchFilter({ [key]: value })),
  setSearchLocation: searchLocation => dispatch(selectedAutocompletionItem(searchLocation)),
  updateSearchPath: searchPath => dispatch(setSearchLocation(searchPath)),
  renderComplete: () => dispatch(renderComplete())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchContainer);
