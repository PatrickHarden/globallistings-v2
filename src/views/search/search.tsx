import React, { FC } from 'react';
import styled from 'styled-components';

import { ButtonInfo, AutocompleteSuggestion } from '../../redux/selectors/search';

import { SearchBias, SearchHeader, SearchPath, Polygon } from '../../types/config';
import { Dictionary } from '../../types/common/dictionary';

import SearchFilters from './search-filters/search-filters';
import SearchHeaderComponent from './search-header/search-header';
import SearchForm from './search-form/search-form';
import { FilterInfo } from '../../redux/selectors/config/filters';

const Container = styled.div``;

interface Props {
  readonly biasRadius?: number;
  readonly buyButtonInfo: ButtonInfo;
  readonly primaryFilters: FilterInfo[];
  readonly secondaryFilters: FilterInfo[];
  readonly language: Dictionary<string>;
  readonly letButtonInfo: ButtonInfo;
  readonly restrictToCountry?: string;
  readonly searchBias?: SearchBias;
  readonly searchHeader: SearchHeader;
  readonly searchMode: string;
  readonly searchResultsPage: string;
  readonly suggestions: AutocompleteSuggestion[];
  readonly changePlaceSearch: (searchTerm: string) => void;
  readonly changePropertySearch: (searchTerm: string) => void;
  readonly changeSearchFilter: (key: string) => (value: string) => void;
  readonly setSearchLocation: (searchLocation: string) => void;
  readonly updateSearchPath: (value: string) => void;
  readonly searchPathSelectors?: SearchPath[];
  readonly searchSuggestDebounce?: number;
  readonly polygons?: Polygon[];
  readonly siteType: string;
}

export const Search: FC<Props> = props => {
  const { changeSearchFilter, primaryFilters, searchHeader, secondaryFilters } = props;

  return (
    <Container>
      <SearchHeaderComponent {...searchHeader} />
      <SearchForm {...props} />
      <SearchFilters
        primaryFilters={primaryFilters}
        secondaryFilters={secondaryFilters}
        handleChange={changeSearchFilter}
      />
    </Container>
  );
};
