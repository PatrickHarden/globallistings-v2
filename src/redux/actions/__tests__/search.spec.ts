import { changeSearchFilter, setSearchLocation } from '../search';

describe(changeSearchFilter, () => {
  it('should create correct action', () => {
    expect(changeSearchFilter({ foo: 'zap' })).toMatchSnapshot();
  });
});

describe(setSearchLocation, () => {
  it('should create correct action', () => {
    expect(setSearchLocation('zap')).toMatchSnapshot();
  });
});
