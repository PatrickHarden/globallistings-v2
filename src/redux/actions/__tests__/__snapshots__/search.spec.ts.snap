// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create correct action 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "foo": "zap",
  },
  "type": "CHANGE_SEARCH_FILTER",
}
`;

exports[`should create correct action 2`] = `
Object {
  "meta": undefined,
  "payload": "zap",
  "type": "SET_SEARCH_LOCATION",
}
`;
